### DISCUSSION TITLE
Combine multiple meta channels into 1 alpha channel

### DISCUSSION THREAD

**Discussion by renderpci** [ImageMagick/ImageMagick#7576] [Help] [answered]
### ImageMagick version

7.1.1-3

### Operating system, version and so on

OSX, Ubuntu 22.04

### Description

Hi

I'm implementing a conversion for different image resolutions and I want to preserve the meta channels as alpha channel in the final image.

The main idea is to create a transparent image with all color channels flat but preserving the meta channel as alpha, in this way:

```
magick  input.tiff -channel-fx "meta=>alpha"  output.avif
```

and it works fine. But when input.tiff has multiple meta channels, how can combine all of them (a Plus blend) into the alpha? 

Thanks in avance.

**Comment by urban-warrior** [comment]
You can reference multiple meta channels as `meta`, `meta1`, `meta2`, etc..  However, as you copy each distinct meta channel to the alpha channel, they will override the previous meta channel.    ImageMagick does not support blending of multiple meta channels into a single alpha channel.

**Reply by fmw42** [reply]
Perhaps export the meta channels to individual images. Then do your blending (composite). Then copy that combined image to an alpha channel.

**Comment by snibgo** [comment, marked as answer]
In theory, something like this (untested, bash syntax):
```
magick \
  in.tiff \
  -alpha opaque \
  \( +clone \
      -set colorspace sRGB \
      -channel-fx "meta0=>red meta1=>green" \
  \) \
  -channel A \
    -fx "v.r+v.g" \
  +channel \
  out.tiff
```
Assuming in.tiff has at least 2 meta channels, this ensures we have an alpha channel, then makes a clone. In that clone we ensure it has RGB channels, and we copy the first two meta channels to red and green.

Now we have two images. Within `-fx`, they are `u` and `v`. This `-fx` will replace both input imsges with one output. It sets the alpha channel of the first image to the sum of the red and green channels of the second image.

If that doesn't work, please post a sample input image.

[Currently, `-fx` can't read meta channels. I may enhance this.]

**Reply by fmw42** [reply]
> "v.r+v.g"

Might that not overflow or clamp if you do not divide by 2?

**Comment by snibgo** [comment]
Yes, just as `-compose Plus -composite` may overflow or clamp. Any expression can be used, such as:
```
 -fx "(v.r+v.g)/2" 
```
My command will create `out.tiff` that contains the original colour channels, and an alpha channel, and the original meta channels. Most image formats can't record meta channels.

**Comment by renderpci** [comment]
Hi!

Thanks snibgo, it works perfectly and I understand the idea; move the meta channels to colour channels and mix... but what happens when the tiff has more than 3 meta channels (or more than 4 if the colour space is CMYK)....

Best

**Comment by snibgo** [comment]
Create as many clones as you want. Copy meta0 to meta2 to the first clone, meta3 to meta5 to the second clone, etc. In the `-fx`, instead of `u` and `v`, you can use `u[0]`, `u[1]`, `u[2]` etc.

I'm curious. Can you say what these images are, with meta channels?

**Comment by renderpci** [comment]
#@snibgo I'm curious. Can you say what these images are, with meta channels?

The images are from cultural heritage (old images made in the 90's), the meta channels were created to remove the background, select objects or areas, the original use was to print into exhibition catalogs and the museums has thousands of them, now we are implementing inside [Dédalo](https://dedalo.dev) a way to create the «final» composition as web version with the same look than the printed catalog has. Dédalo store the original tiff, but the idea is create automatically an .avif version with alpha channel with all meta channels applied.

Thanks again for your help.

Best
